File: sdk/js/node/src/permissions.ts
import type { HttpClient } from "./client.js";
import type { ListParams, Page } from "./users.js";

export interface Permission {
  id: string;
  name: string;
  description?: string;
  created_at: string;
  updated_at?: string;
}

export interface CreatePermissionInput {
  name: string;
  description?: string;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
  next_cursor?: string;
}

export class Permissions {
  constructor(private readonly http: HttpClient) {}

  create(input: CreatePermissionInput): Promise<Permission> {
    return this.http.post<Permission>("/v1/rbac/permissions", input);
  }

  get(id: string): Promise<Permission> {
    return this.http.get<Permission>(`/v1/rbac/permissions/${encodeURIComponent(id)}`);
  }

  delete(id: string): Promise<void> {
    return this.http.delete<void>(`/v1/rbac/permissions/${encodeURIComponent(id)}`);
  }

  async list(params: ListParams = {}): Promise<Page<Permission>> {
    const res = await this.http.get<ListEnvelope<Permission>>("/v1/rbac/permissions", {
      query: { tenant: params.tenant, limit: params.limit, cursor: params.cursor },
    });
    return { data: res.items ?? res.data ?? [], nextCursor: res.next_cursor };
  }

  async *listAll(params: ListParams = {}): AsyncGenerator<Permission> {
    let cursor = params.cursor;
    do {
      const page = await this.list({ ...params, cursor });
      for (const item of page.data) yield item;
      cursor = page.nextCursor;
    } while (cursor);
  }
}

File: sdk/go/permissions.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

type Permission struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at,omitempty"`
}

type CreatePermissionInput struct {
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
}

type PermissionPage struct {
	Data       []Permission
	NextCursor string
}

type Permissions struct{ http *httpClient }

func (r *Permissions) Create(ctx context.Context, in CreatePermissionInput) (*Permission, error) {
	var out Permission
	err := r.http.do(ctx, http.MethodPost, "/v1/rbac/permissions", nil, in, &out, false)
	return &out, err
}

func (r *Permissions) Get(ctx context.Context, id string) (*Permission, error) {
	var out Permission
	err := r.http.do(ctx, http.MethodGet, "/v1/rbac/permissions/"+url.PathEscape(id), nil, nil, &out, true)
	return &out, err
}

func (r *Permissions) Delete(ctx context.Context, id string) error {
	return r.http.do(ctx, http.MethodDelete, "/v1/rbac/permissions/"+url.PathEscape(id), nil, nil, nil, true)
}

func (r *Permissions) List(ctx context.Context, params ListParams) (*PermissionPage, error) {
	q := url.Values{}
	if params.Tenant != "" {
		q.Set("tenant", params.Tenant)
	}
	if params.Limit > 0 {
		q.Set("limit", strconv.Itoa(params.Limit))
	}
	if params.Cursor != "" {
		q.Set("cursor", params.Cursor)
	}
	var env struct {
		Items      []Permission `json:"items"`
		Data       []Permission `json:"data"`
		NextCursor string       `json:"next_cursor"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/rbac/permissions", q, nil, &env, true); err != nil {
		return nil, err
	}
	items := env.Items
	if items == nil {
		items = env.Data
	}
	return &PermissionPage{Data: items, NextCursor: env.NextCursor}, nil
}
